File: 13-1.py
#!/usr/bin/env python

def compare(p1, p2):
    if len(p1) == 0 and len(p2) == 0:
        return 0
    if len(p1) == 0:
        return -1
    if len(p2) == 0:
        return 1

    left = p1[0]
    right = p2[0]

    if type(left) is int and type(right) is int:
        if left == right:
            return compare(p1[1:], p2[1:])
        elif left < right:
            return -1
        else:
            return 1
    elif type(left) is list and type(right) is list:
        r = compare(left, right)
        if r == 0:
            return compare(p1[1:], p2[1:])
        else:
            return r
    elif type(left) is int:
        r = compare([left], right)
        if r == 0:
            return compare(p1[1:], p2[1:])
        else:
            return r
    elif type(right) is int:
        r = compare(left, [right])
        if r == 0:
            return compare(p1[1:], p2[1:])
        else:
            return r
    else:
        raise RuntimeError(f"Type error: {left}, {right}")


input = [row.strip() for row in open('13.input').readlines()]

score = 0
index = 0

while len(input) > 0:
    index += 1
    p1 = eval(input[0])
    p2 = eval(input[1])

    if compare(p1, p2) == -1:
        score += index

    input = input[2:]
    if len(input) > 0 and input[0] == "":
        input = input[1:]

print(score)

File: 13-1.rb
#!/usr/bin/env ruby

def compare(p1, p2)
  left = p1.shift
  right = p2.shift

  return :equal if left.nil? and right.nil?
  return :less if left.nil?
  return :greater if right.nil?

  if left.is_a? Integer and right.is_a? Integer
    if left == right
      compare(p1, p2)
    elsif left < right
      :less
    else
      :greater
    end
  elsif left.is_a? Array and right.is_a? Array
    case compare(left, right)
    when :equal
      compare(p1, p2)
    when :less
      :less
    when :greater
      :greater
    end
  elsif left.is_a? Integer
    case compare([left], right)
    when :equal
      compare(p1, p2)
    when :less
      :less
    when :greater
      :greater
    end
  elsif right.is_a? Integer
    case compare(left, [right])
    when :equal
      compare(p1, p2)
    when :less
      :less
    when :greater
      :greater
    end
  else
    raise "Type error: #{left}, #{right}"
  end
end

input = File.read('13.input').lines.map(&:strip)

score = 0
index = 0

until input.empty?
  index += 1
  p1 = eval(input.shift)
  p2 = eval(input.shift)

  score += index if compare(p1, p2) == :less

  input.shift if input[0] == ''
end

print score, "\n"
